dataRetriever/dataPool/proofsCache/proofsCache.go

```diff
@@ -43,39 +46,76 @@ func (pc *proofsCache) addProof(proof data.HeaderProofHandler) {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.insertProofByNonce(proof)
 
-	pc.proofsByNonce = append(pc.proofsByNonce, &proofNonceMapping{
-		headerHash: string(proof.GetHeaderHash()),
-		nonce:      proof.GetHeaderNonce(),
-	})
+	pc.mutProofsByHash.Lock()
+	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	pc.mutProofsByHash.Unlock()
+}
 
-	sort.Slice(pc.proofsByNonce, func(i, j int) bool {
-		return pc.proofsByNonce[i].nonce < pc.proofsByNonce[j].nonce
-	})
+func (pc *proofsCache) insertProofByNonce(proof data.HeaderProofHandler) {
+	pc.mutProofsByNonce.Lock()
+	defer pc.mutProofsByNonce.Unlock()
 
-	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	if len(pc.proofsByNonceBuckets) == 0 {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket := pc.proofsByNonceBuckets[0]
+
+	if headBucket.isFull() {
+		pc.insertInNewBucket(proof)
+		return
+	}
+
+	headBucket.insertInExisting(proof)
+}
+
+func (pc *proofsCache) insertInNewBucket(proof data.HeaderProofHandler) {
+	bucket := newProofBucket(pc.bucketSize)
+	bucket.insertInNew(proof)
+
+	pc.proofsByNonceBuckets = append([]*proofNonceBucket{bucket}, pc.proofsByNonceBuckets...)
 }
 
 func (pc *proofsCache) cleanupProofsBehindNonce(nonce uint64) {
 	if nonce == 0 {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.mutProofsByNonce.Lock()
+	defer pc.mutProofsByNonce.Unlock()
+
+	buckets := make([]*proofNonceBucket, 0)
+
+	wg := &sync.WaitGroup{}
+
+	for _, bucket := range pc.proofsByNonceBuckets {
+		if nonce > bucket.maxNonce {
+			wg.Add(1)
 
-	proofsByNonce := make([]*proofNonceMapping, 0)
+			go func(bucket *proofNonceBucket) {
+				pc.cleanupProofsInBucket(bucket)
+				wg.Done()
+			}(bucket)
```

Comment:
Maybe allow the cleanup to happen synchronously? I see the workload is _lock, delete from map in a loop, unlock_. Do we have a speed gain using concurrency in this context (since map deletions are not quite subject to parallelization)?

Reply:
i did it initially without concurrency, but after i did some benchmarks, there seems to be a 2x improvement; but i agree, not sure if it's worth it

Reply:
removed concurrency here